returnn/util/sig_proc.py

```diff
@@ -0,0 +1,159 @@
+
+  def get_gammatone_filterbank(self):
+    """
+    Returns an array with the parameters of the gammatone filterbank
+
+    :return: gammatone filterbank
+    :rtype: numpy.array
+    """
+    center_freqs = self.center_frequencies(self.num_channels, self.freq_max, self.freq_min)
+    fbank = []
+    for freq in center_freqs:
+      fbank.append(self.gammatone_impulse_response(freq, self.length, self.sample_rate))
+    fbank = numpy.vstack(fbank)
+    if self.normalization:
+      fbank = self.normalize_filters(fbank)
+    fbank = numpy.transpose(fbank)
+    return fbank
+
+  @staticmethod
+  def center_frequencies(num_channels, freq_max, freq_min):
+    """
+    Determine center frequencies for gammatone filterbank
+
+    :param int num_channels: number of filters
+    :param float freq_max: maximum frequency of filterbank
+    :param float freq_min: minimum frequency of filterbank
+    :return: center frequencies
+    :rtype: numpy.array
+    """
+    x_min = inv_greenwood_function(freq_min)
+    x_max = inv_greenwood_function(freq_max)
+    x_linear = numpy.linspace(x_min, x_max, num_channels)
+    freq_center = numpy.vectorize(greenwood_function)(x_linear)
+    return freq_center
+
+  @staticmethod
+  def bandwidth_by_center_frequency(freq, l_=24.7, q_=9.264491981582191):
+    """
+    Get bandwidth (named B in [1]) by center frequency using a linear approximation of the equivalent rectangular
+    bandwidth from
+    Glasberg, Brian R., and Brian CJ Moore. "Derivation of auditory filter shapes from notched-noise data."
+    Hearing research, 1990
+    The values are taken from there and are also used in RASR, see
+    https://github.com/rwth-i6/rasr/blob/master/src/Signal/GammaTone.cc
+
+    :param float freq: center frequency
+    :param float l_:
+    :param float q_: audiological (ERB) based filter quality factor
+    :return: bandwidth
+    :rtype: float
+    """
+    return l_ * (1 / (l_ * q_) * freq + 1.0)
+
+  def gammatone_impulse_response(self, f_center, length, sample_rate, output_gain=1., filter_order=4, phase_shift=0.):
+    """
+    Compute gammatone impulse response based on [1]
+
+    :param float f_center: center frequency
+    :param int length: length of finite impulse response in seconds
+    :param int sample_rate: sample rate of audio signal in Hz
+    :param float output_gain: output gain, named k in [1]
+    :param int filter_order: order of filter, named n in [1]
+    :param float phase_shift: phase shift, named phi in [1]
+    :return: gammatone impulse response
+    :rtype: numpy.array
+    """
+    bandwidth = self.bandwidth_by_center_frequency(f_center)  # bandwidth (duration of impulse response)
+    num_samples = int(numpy.floor(sample_rate * length))
+    t = numpy.linspace(1. / sample_rate, length, num_samples)
+    return output_gain * numpy.power(t, filter_order - 1) * numpy.exp(-2 * numpy.pi * bandwidth * t) * (
+      numpy.cos(2 * numpy.pi * f_center * t + phase_shift))
```

Comment:
You should not indent one part of it and not the other. Either all or nothing.
```suggestion
    return (
      output_gain *
      numpy.power(t, filter_order - 1) *
      numpy.exp(-2 * numpy.pi * bandwidth * t) *
      numpy.cos(2 * numpy.pi * f_center * t + phase_shift))
```